### PR TITLE
Add SQL error guidance to developer style guide

### PR DESCRIPTION
**Pull request by chaas**
Introduce formal SQL error guide to our developer docs, based on PostgreSQL's official manual ( #15630 ). 

Thoughts on whether we should include links to our docs in hints?

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by chaas** [comment]
Using https://github.com/MaterializeInc/materialize/issues/15502 as an example, it would look like
```
ERROR:  type with OID 211538 is unknown
DETAIL:  Unrecognized type for column "current_mood" in table "person".
HINT:  You may be using an unsupported type in Materialize, such as an enum.
```

**Review by ggnall — COMMENTED** [review]
Style guide looks good! I think we should make "Hints" explicit directions of what to do next. 

To nitpick the example - We should tell a user how to remove or modify the unsupported type. Like: "Try DROPing your Connection, casting this column to a string, and re-CREATEing the source" I made up that solution but hopefully it's illustrative about how we should use specifics.

> HINT:  You may be using an unsupported type in Materialize, such as an enum.

File: doc/developer/style.md [hunk 1]
```diff
@@ -92,6 +92,28 @@ not entirely consistent on the finer points, so don't worry about being exact.
 But please at least capitalize the obvious keywords, like `SELECT`, `INSERT`,
 `FROM`, `WHERE`, `AS`, `AND`, etc.
 
+### SQL error style
+We adhere to the PostgreSQL manual's [SQL error message style guide](https://www.postgresql.org/docs/current/error-style-guide.html).
+
+The format consists of a required primary message and optional detail and hint messages.
```
Comment (morsapaes) [inline doc/developer/style.md:98]:
```suggestion
For error message style, we strive to adhere to the [guidelines](https://www.postgresql.org/docs/current/error-style-guide.html) in the PostgreSQL manual, and use a format that consists of a **required** _primary_ message followed by **optional** _detail_ and _hint_ messages.
```

**Review by morsapaes — COMMENTED** [review]
Thanks for pinging me, @chaas! 

Since chunks of this seem to be taken verbatim from the PostgreSQL documentation, do we need to enclose the text in quotes, or [mention the license](https://www.postgresql.org/about/licence/) somewhere? Copyrighting makes me hyperventilate — so just double checking! It could be easier to just slightly rephrase this using our own words.

> Thoughts on whether we should include links to our docs in hints?

Documentation links might get stale pretty fast (though we can always alias ), at the pace we move things around, so I'd say 👎.

**Comment by benesch** [comment]
> We could change the first line to type with OID 211538 is unrecognized too if we're ok changing the existing error string that's used in a bunch of places. I could also just bump the details up into the error?

Yeah, def okay with changing the error! I feel like the error message and detail should perhaps be reversed?

```
ERROR:  column "person.current_mood" uses unrecognized type
DETAIL:  The column type has OID 211538.
HINT:  You may be using an unsupported type in Materialize, such as an enum.
```

> Since chunks of this seem to be taken verbatim from the PostgreSQL documentation, do we need to enclose the text in quotes, or [mention the license](https://www.postgresql.org/about/licence/) somewhere? Copyrighting makes me hyperventilate — so just double checking! It could be easier to just slightly rephrase this using our own words.

We've already got a copy of the license in our LICENSE file, so just need to drop a note in the document like "Portions of this section are copied from the PostgreSQL documentation and subject to the PostgreSQL license, a copy of which is available in the LICENSE file at the root of this repository."

> Style guide looks good! I think we should make "Hints" explicit directions of what to do next.
> 
> To nitpick the example - We should tell a user how to remove or modify the unsupported type. Like: "Try DROPing your Connection, casting this column to a string, and re-CREATEing the source" I made up that solution but hopefully it's illustrative about how we should use specifics.

> > Thoughts on whether we should include links to our docs in hints?
> 
> Documentation links might get stale pretty fast (though we can always alias pages as we go), at the pace we move things around, so I'd say 👎.

It's often hard to cram all that advice into the `HINT` field. A sentence or two works well, but often you want to show a code sample etc. and that quickly overwhelms what the HINT field is capable of formatting. CockroachDB links to docs in error messages and it works _really_ well.

My recommendation is that we use shortlinks for this! We've got an example of this for our old non-materialized error that was hard to explain:

https://github.com/MaterializeInc/materialize/blob/3efb2c933e4e6c8e694afb1e794c7d8dae368e7d/ci/deploy/website.sh#L31

The idea is that once you introduce a shortlink (e.g., https://materialize.com/s/unknown-postgres-type), we're committed to the upkeep of that link. And since the shortlinks are defined in this repo, it's easy to add the shortlink in the same PR that you reference the shortlink in an error message.

**Comment by morsapaes** [comment]
> The idea is that once you introduce a shortlink (e.g., https://materialize.com/s/unknown-postgres-type), we're committed to the upkeep of that link. And since the shortlinks are defined in this repo, it's easy to add the shortlink in the same PR that you reference the shortlink in an error message.

This sounds great, @benesch!

**Comment by chaas** [comment]
Thanks for the feedback!

>  Copyrighting makes me hyperventilate 

Good call! I'm not used to having my internal developer docs be open source 😅. I'll make sure to add the right disclosures/not lift the guide so directly.

> My recommendation is that we use shortlinks for this!

Very cool that we already support this - sounds good.

>  I feel like the error message and detail should perhaps be reversed?
Good suggestion. We use the "type with OID x is unknown" error message in numerous other places too, so perhaps we should follow up to change all of those cases to be more descriptive.

The one thing I'd like to clarify further is what's deserving of the Error message vs the Detail. The Postgres guide is fairly light on details or examples, and we haven't added any more rationale either.

**Review by morsapaes — APPROVED** [review]

**Review by benesch — APPROVED** [review]

**Pull request merged**